elasticsearch/class-health-command.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace Automattic\VIP\Elasticsearch;
+
+use \WP_CLI;
+use \WP_CLI\Utils;
+use \WP_Query as WP_Query;
+use \ElasticPress\Indexables as Indexables;
+use \ElasticPress\Elasticsearch as Elasticsearch;
+
+/**
+ * Helper commands to manage VIP Go Elasticsearch indexes
+ *
+ * @package Automattic\VIP\Elasticsearch
+ */
+class Health_Command extends \WPCOM_VIP_CLI_Command {
+
+	/**
+	 * @var WP_CLI\Fetchers\Plugin Plugin fetcher
+	 */
+	protected $plugin_fetcher;
+
+	public function __construct() {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		parent::__construct();
+
+		$this->plugin_fetcher = new \WP_CLI\Fetchers\Plugin();
+	}
+
+	/**
+	 * Validate DB and ES index post counts
+	 *
+	 * ## OPTIONS
+	 *
+	 *
+	 * ## EXAMPLES
+	 *     wp vip-es health validate-counts
+	 *
+	 * @subcommand validate-counts
+	 */
+	public function validate_counts( $args, $assoc_args ) {
+		// Make sure ElasticPress is installed and active
+		$this->is_elasticpress_active();
+
+		// Get indexable objects
+		$indexable = Indexables::factory()->get( 'post' );
+
+		$post_types = $indexable->get_indexable_post_types();
+
+		WP_CLI::line( sprintf( "Checking %d post types (%s)\n", count( $post_types ), implode( ',', $post_types ) ) );
+
+		$error = false;
+		$es_conn_err = false;
+
+		foreach( $post_types as $post_type ) {
+			$query_args = [ 
+				'post_type' => $post_type,
+			];
+
+			// Get total count in DB
+			$result = $indexable->query_db( $query_args );
+
+			$db_total = (int) $result[ 'total_objects' ];
+
+			// Get total count in ES index
+			$query = new WP_Query( $query_args );
+
+			$formatted_args = $indexable->format_args( $query->query_vars, $query );
+
+			$es_result = $indexable->query_es( $formatted_args, $query->query_vars );
+			
+			if ( ! $es_result ) {
+				$es_total = 'N/A';
+				$error = true;
+				$es_conn_err = true;
+			} else {
+				$es_total = (int) $es_result[ 'found_documents' ][ 'value' ];
+			}
+
+			$icon = "\u{2705}"; // unicode check mark
+			$diff = '';
+			if ( $db_total !== $es_total ) {
+				$error = true;
+
+				$icon = "\u{274C}"; // unicode cross mark
+
+				if ( $es_result ) {
+					$diff = sprintf( ', diff: %d', $es_total - $db_total );
```

Comment:
The DB could also have more items than the ES index, which would cause this to display a negative number.

Reply:
That is deliberate. A negative diff would indicate the ES index is missing items while a positive diff would mean the opposite

Reply:
That might be confusing to read as someone who runs the command, because you may not know which is the "left" side vs "right" side of the equation (so without knowledge of the code, you don't know if negative means posts missing from the DB or from the index).

For clarity, I'd suggest a more explicit message, like `index has 18 more posts than db` and `db has 90 more posts than index`.

Reply:
Thoughts on shipping as-is, and fixing up the output to be nicer in a future PR?

Reply:
Ya that would be fine, it's minor.